src/main/java/net/snowflake/client/jdbc/SnowflakeSQLLoggedException.java

```diff
@@ -0,0 +1,267 @@
+    return ibInstance.sendBatchAsync();
+  }
+
+  /**
+   * Helper function to create JSONObject node for OOB telemetry log
+   * @param queryId
+   * @param reason
+   * @param SQLState
+   * @param vendorCode
+   * @param errorCode
+   * @return JSONObject with data about SQLException
+   */
+  private JSONObject createOOBValue(String queryId, String reason, String SQLState, int vendorCode, ErrorCode errorCode)
+  {
+    JSONObject oobValue = new JSONObject();
+    if (!Strings.isNullOrEmpty(queryId))
+    {
+      oobValue.put("Query ID", queryId);
+    }
+    if (!Strings.isNullOrEmpty(SQLState))
+    {
+      oobValue.put("SQLState", SQLState);
+    }
+    if (vendorCode != -1)
+    {
+      oobValue.put("Vendor Code", vendorCode);
+    }
+    if (errorCode != null)
+    {
+      oobValue.put("Error Code", errorCode.toString());
+    }
+    if (!Strings.isNullOrEmpty(reason))
+    {
+      oobValue.put("reason", reason);
+    }
+    return oobValue;
+  }
+
+  /**
+   * Function to construct log data based on possible exception inputs and send data through in-band telemetry, or oob
+   * if in-band does not work
+   * @param queryId query ID if exists
+   * @param reason reason for the exception
+   * @param SQLState SQLState
+   * @param vendorCode vendor code
+   * @param errorCode error code
+   * @param session session object, which is needed to send in-band telemetry but not oob. Might be null, in which case
+   *                oob is used.
+   */
+  public void sendTelemetryData(String queryId, String reason, String SQLState, int vendorCode, ErrorCode errorCode, SFSession session)
+  {
+    // if session is not null, try sending data using in-band telemetry
+    if (session != null)
+    {
+      ibInstance = session.getTelemetryClient();
+    }
+    // if in-band instance is successfully created, compile sql exception data into an in-band telemetry log
+    if (ibInstance != null)
+    {
+      ObjectNode ibValue = mapper.createObjectNode();
+      ibValue.put("type", TelemetryField.SQL_EXCEPTION.toString());
+      if (!Strings.isNullOrEmpty(queryId))
+      {
+        ibValue.put("Query ID", queryId);
+      }
+      if (!Strings.isNullOrEmpty(SQLState))
+      {
+        ibValue.put("SQLState", SQLState);
+      }
+      if (vendorCode != -1)
+      {
+        ibValue.put("Vendor Code", vendorCode);
+      }
+      if (errorCode != null)
+      {
+        ibValue.put("Error Code", errorCode.toString());
+      }
+      if (!Strings.isNullOrEmpty(reason))
+      {
+        ibValue.put("reason", reason);
+      }
+      // try  to send in-band data asynchronously
+      ExecutorService threadExecutor = Executors.newSingleThreadExecutor();
+      threadExecutor.submit(() ->
+        {
+          boolean inBandSuccess;
+          Future<Boolean> sendInBand = sendInBandTelemetryMessage(ibValue, this);
+          // record whether in band telemetry message sent with boolean value inBandSuccess
+          try
```

Comment:
Try/catch is needed after all because Future.get() can throw several types of exceptions